src/compiler/crystal/semantic/call_error.cr

```diff
@@ -194,6 +176,311 @@ class Crystal::Call
     raise message, owner_trace
   end
 
+  private def check_block_mismatch(call_errors, owner, def_name)
+    return unless call_errors.all?(BlockMismatch)
+
+    error_message =
+      if block
+        "'#{full_name(owner, def_name)}' is not expected to be invoked with a block, but a block was given"
+      else
+        "'#{full_name(owner, def_name)}' is expected to be invoked with a block, but no block was given"
+      end
+
+    raise error_message
+  end
+
+  private def check_missing_named_arguments(call_errors, owner, defs, arg_types, inner_exception)
+    return unless call_errors.all?(MissingNamedArguments)
+
+    call_errors = call_errors.map &.as(MissingNamedArguments)
+    all_missing_names = call_errors.flat_map(&.names).uniq
+    missing_names_in_all_overloads = all_missing_names.select do |missing_name|
```

Comment:
Minor performance improvement:
```suggestion
    all_missing_names = call_errors.flat_map(&.names).uniq!
    missing_names_in_all_overloads = all_missing_names.select do |missing_name|
```

Technically the `select` call could also be mutating, but it's stored in a different variable (even though it's never used again, so there is no side effect).

The same applies to the similar code sections in the following methods. Maybe it would be a good idea to extract this tally logic into a helper method? It seems to do pretty much the same thing.

Reply:
I'll try to, even though the code is similar it's not exactly the same, and extracting a common method might make things harder to understand. But I'll try!

Reply:
Done! Only one of them remains different now, but because it's different (it's the most complex case, the one about argument type mismatch)